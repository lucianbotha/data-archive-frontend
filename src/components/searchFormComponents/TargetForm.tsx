import * as React from "react";
import targetPosition from "target-position";
import { ITarget } from "../../utils/ObservationQueryParameters";
import {
  isFloat,
  validateDec,
  validateName,
  validateRa
} from "../../utils/validators";
import { InnerMainGrid, MainGrid, SubGrid } from "../basicComponents/Grids";
import InputField from "../basicComponents/InputField";
import SelectField from "../basicComponents/SelectField";

class TargetForm extends React.Component<
  { target: ITarget; onChange: any },
  any
> {
  resolve = () => {
    const { target, onChange } = this.props;
    const resolver = target.resolver || "Simbad";

    targetPosition(target.name || "", [resolver])
      .then(p => {
        if (p) {
          onChange({
            ...target,
            declination: `${p.declination}`,
            rightAscension: `${p.rightAscension}`
          });
        } else {
          onChange({
            ...target,
            errors: {
              ...target.errors,
              name: `The target name ${target.name} could not be resolved.`
            }
          });
        }
      })
      .catch(err => {
        onChange({
          ...target,
          errors: {
            ...target.errors,
            name: err.message
          }
        });
      });
  };
  render() {
    const { target, onChange } = this.props;
    const targetChange = (
      e: React.FormEvent<HTMLSelectElement | HTMLInputElement>
    ) => {
      const name = e.currentTarget.name;
      const value = e.currentTarget.value;
      onChange({
        ...target,
        [name]: value,
        errors: {
          ...target.errors,
          [name]: ""
        }
      });
    };
    return (
      <>
        <MainGrid>
          <SubGrid>
            <p>Target name</p>
            <InputField
              name={"name"}
              value={target.name || ""}
              error={target.errors.name || ""}
              onChange={targetChange}
            />
          </SubGrid>
          <SubGrid>
            <InnerMainGrid>
              <SubGrid>
                <p>Resolver</p>
                <SelectField
                  options={["Simbad", "NED", "VizieR"]}
                  name={"resolver"}
                  value={target.resolver || "Simbad"}
                  onChange={targetChange}
                />
              </SubGrid>
              <SubGrid>
                <br />
                <input
                  className="button is-info"
                  type="button"
                  value="resolve"
                  onClick={this.resolve}
                />
              </SubGrid>
            </InnerMainGrid>
          </SubGrid>
        </MainGrid>

        <MainGrid>
          <SubGrid>
            <p>Right ascension</p>
            <InputField
              name={"rightAscension"}
              value={target.rightAscension || ""}
              onChange={targetChange}
              error={target.errors.rightAscension}
            />
          </SubGrid>
          <SubGrid>
            <p>Declination</p>
            <InputField
              name={"declination"}
              value={target.declination || ""}
              onChange={targetChange}
              error={target.errors.declination}
            />
          </SubGrid>
        </MainGrid>
        <MainGrid>
          <SubGrid>
            <p>Search radius</p>
            <InputField
              name={"searchConeRadius"}
              value={target.searchConeRadius || ""}
              onChange={targetChange}
              error={target.errors.searchConeRadius}
            />
          </SubGrid>
          <SubGrid>
            <p>Radius units</p>
            <SelectField
              options={["Arcseconds", "Arcminutes", "Degrees"]}
              name={"radiusUnits"}
              onChange={targetChange}
              value={target.searchConeRadiusUnits || "Arcseconds"}
            />
          </SubGrid>
        </MainGrid>
      </>
    );
  }
}

/**
 */
export const validatedTarget = async (target: ITarget) => {
  let raDecChangeError;
  if (target.name && target.name !== "") {
    raDecChangeError = await targetPosition(target.name, [
      target.resolver || "Simbad"
    ])
      .then(p => {
        if (p) {
          return {
            declination:
              `${p.declination}` !== target.declination
                ? `The declination differs from that returned by ${
                    target.resolver
                  } for ${target.name}.`
                : "",
            rightAscension:
              `${p.rightAscension}` !== target.rightAscension
                ? `The right ascension differs from that returned by ${
                    target.resolver
                  } for ${target.name}.`
                : ""
          };
        }
      })
      .catch();
  }
  return {
    ...target,
    errors: {
      declination:
        raDecChangeError && raDecChangeError.declination
          ? raDecChangeError.declination
          : validateDec(target.declination || ""),
      name: validateName(target.name || ""),
      rightAscension:
        raDecChangeError && raDecChangeError.rightAscension
          ? raDecChangeError.rightAscension
          : validateRa(target.rightAscension || ""),
      searchConeRadius: isFloat(target.searchConeRadius || "")
        ? ""
        : "The search radius must be a positive number."
    }
  };
};
export default TargetForm;
