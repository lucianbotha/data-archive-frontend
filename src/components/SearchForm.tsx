import * as React from "react";
import {
  CalibrationType,
  IGeneral,
  ISearchFormState,
  ITarget,
  ITelescope
} from "../utils/ObservationQueryParameters";
import {
  ButtonGrid,
  DataGrid,
  ParentGrid,
  ProposalGrid,
  Spinner,
  TargetGrid,
  TelescopeGrid
} from "./basicComponents/Grids";
import DataForm from "./searchFormComponents/DataForm";
import ProposalForm, {
  validatedProposal
} from "./searchFormComponents/ProposalForm";
import TargetForm, { validatedTarget } from "./searchFormComponents/TargetForm";
import TelescopeForm, {
  validatedTelescope
} from "./searchFormComponents/TelescopeForm";

/**
 * A form for defining search parameters for an observation search, and for
 * initiating the search.
 */
class SearchForm extends React.Component<{}, ISearchFormState> {
  public state: ISearchFormState = {
    general: { calibrations: new Set<CalibrationType>(), errors: {} },
    loading: false,
    target: {
      errors: {},
      resolver: "Simbad",
      searchConeRadiusUnits: "arcseconds"
    }
  };

  /**
   * Handle changes of telescope-related parameters.
   */
  public telescopeChange = (value: ITelescope) => {
    const newState = {
      ...this.state,
      telescope: {
        ...value
      }
    };
    this.setState(() => newState);
  };

  /**
   * Handle changes of target-related parameters.
   */
  public targetChange = (value: ITarget) => {
    const newState = {
      ...this.state,
      target: {
        ...value
      }
    };
    this.setState(() => newState);
  };

  /**
   * Handle changes of general parameters.
   */
  public generalChange = (value: IGeneral) => {
    const newState = {
      ...this.state,
      general: {
        ...value
      }
    };
    this.setState(() => newState);
  };

  /**
   * Perform an observation with the currently selected search parameters.
   */
  public searchArchive = async () => {
    this.setState(() => ({
      ...this.state,
      loading: true
    }));

    // Add errors to the search parameter details
    const target = await validatedTarget(this.state.target);
    const general = validatedProposal(this.state.general);
    const telescope = validatedTelescope(this.state.telescope);

    this.setState(() => ({
      ...this.state,
      general,
      loading: false,
      target,
      telescope
    }));
  };

  public render() {
    const { target, general, telescope, loading } = this.state;

    return (
      <>
        {loading && (
          <div className="dimScreen">
            <Spinner />
          </div>
        )}
        <ParentGrid>
          <TargetGrid>
            <TargetForm target={target} onChange={this.targetChange} />
          </TargetGrid>
          <ProposalGrid>
            <ProposalForm proposal={general} onChange={this.generalChange} />
          </ProposalGrid>
          <TelescopeGrid>
            <TelescopeForm
              telescope={telescope}
              onChange={this.telescopeChange}
            />
          </TelescopeGrid>
          <DataGrid>
            <DataForm general={general} onChange={this.generalChange} />
          </DataGrid>
          <ButtonGrid>
            <input
              className="button is-primary"
              type="button"
              value="Search"
              onClick={this.searchArchive}
            />
          </ButtonGrid>
        </ParentGrid>
      </>
    );
  }
}

export default SearchForm;
