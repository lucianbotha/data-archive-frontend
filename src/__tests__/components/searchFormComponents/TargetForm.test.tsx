import { mount, shallow } from "enzyme";
import toJson from "enzyme-to-json";
import * as React from "react";
import SearchForm from "../../../components/SearchForm";
import TargetForm from "../../../components/searchFormComponents/TargetForm";

const onChange = jest.fn();

describe("TargetForm", () => {
  const wrapper = mount(
    <TargetForm
      target={{
        errors: {},
        resolver: "Simbad",
        searchConeRadiusUnits: "arcseconds"
      }}
      onChange={onChange}
    />
  );

  it("should render", () => {
    expect(wrapper).toBeDefined();
  });

  it("should render correctly", () => {
    expect(
      toJson(
        shallow(
          <TargetForm
            target={{
              errors: {},
              resolver: "Simbad",
              searchConeRadiusUnits: "arcseconds"
            }}
            onChange={onChange}
          />
        )
      )
    ).toMatchSnapshot();

    expect(
      toJson(
        shallow(
          <TargetForm
            target={{
              errors: {},
              name: "NGC 1234",
              resolver: "Simbad",
              searchConeRadiusUnits: "arcseconds"
            }}
            onChange={onChange}
          />
        )
      )
    ).toMatchSnapshot();

    expect(
      toJson(
        shallow(
          <TargetForm
            target={{
              errors: {
                name: "this is an error"
              },
              resolver: "Simbad",
              searchConeRadius: "12.1",
              searchConeRadiusUnits: "arcminutes"
            }}
            onChange={onChange}
          />
        )
      )
    ).toMatchSnapshot();
  });
});

describe("Target form on change", () => {
  it("should find target change defined on target search form", () => {
    // jest.spyOn(SearchForm.prototype, 'targetChange');
    SearchForm.prototype.targetChange = jest.fn();
    expect(SearchForm.prototype.targetChange).toBeDefined();
  });
  const wrapper = mount(
    <TargetForm
      target={{
        errors: {},
        resolver: "Simbad",
        searchConeRadiusUnits: "arcseconds"
      }}
      onChange={onChange}
    />
  );
  it("should call onChange with targetName", () => {
    const target = wrapper.find("input.target-name-input");
    target.simulate("change", "");
    expect(onChange).toBeCalled();
  });
  it("should call onChange with right ascension", () => {
    let value: any;
    let event: any;
    const resolverForm = wrapper.find("input.right-ascension-input");
    const resolver = resolverForm.find("input");
    value = "123.1";
    event = { target: { value, name: "rightAscension" } };
    resolver.simulate("change", event);
    expect(onChange).toBeCalledWith({
      errors: { rightAscension: "" },
      rightAscension: "123.1"
    });

    value = "";
    event = { target: { value, name: "rightAscension" } };
    resolver.simulate("change", event);
    expect(onChange).toBeCalledWith({
      errors: { rightAscension: "" },
      rightAscension: ""
    });

    value = undefined;
    event = { target: { value, name: "rightAscension" } };
    resolver.simulate("change", event);
    expect(onChange).toBeCalledWith({
      errors: { rightAscension: "" },
      rightAscension: ""
    });
  });

  it("should call onChange for selectors as well", () => {
    const resolver = wrapper.find("div.resolver-select");
    resolver.simulate("change", "");
    expect(onChange).toBeCalled();

    const rUnits = wrapper.find("div.radius-units-select");
    rUnits.simulate("change", "");
    expect(onChange).toBeCalled();
  });
});
