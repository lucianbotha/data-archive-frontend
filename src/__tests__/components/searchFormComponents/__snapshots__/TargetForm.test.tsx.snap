// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TargetForm should render correctly 1`] = `
<Fragment>
  <styled.div>
    <styled.div>
      <p>
        Target name
      </p>
      <InputField
<<<<<<< HEAD:src/__tests__/components/searchFormComponents/__snapshots__/TargetForm.test.tsx.snap
=======
        className="target-name-input"
>>>>>>> b25246a6a08e483ea3beb3773509373aaefa6819:src/__tests__/components/searchFormComponents/__snapshots__/TargetForm.test.tsx.snap
        data-test="target-name-input"
        disabled={false}
        name="name"
        onChange={[Function]}
      />
    </styled.div>
    <styled.div>
      <styled.div>
        <styled.div>
          <p>
            Resolver
          </p>
          <SelectField
<<<<<<< HEAD:src/__tests__/components/searchFormComponents/__snapshots__/TargetForm.test.tsx.snap
=======
            className="resolver-select"
>>>>>>> b25246a6a08e483ea3beb3773509373aaefa6819:src/__tests__/components/searchFormComponents/__snapshots__/TargetForm.test.tsx.snap
            data-test="resolver-select"
            name="resolver"
            onChange={[Function]}
            value="Simbad"
          >
            <option
              value="Simbad"
            >
              Simbad
            </option>
            <option
              value="NED"
            >
              NED
            </option>
            <option
              value="VizieR"
            >
              VizieR
            </option>
          </SelectField>
        </styled.div>
        <styled.div>
          <br />
          <button
            className="button is-info false"
            disabled={false}
            onClick={[Function]}
            title=""
            type="button"
          >
            resolve
          </button>
        </styled.div>
      </styled.div>
    </styled.div>
  </styled.div>
  <styled.div>
    <styled.div>
      <p>
        Right ascension
      </p>
      <InputField
        data-test="right-ascension-input"
        disabled={false}
        name="rightAscension"
        onChange={[Function]}
      />
    </styled.div>
    <styled.div>
      <p>
        Declination
      </p>
      <InputField
        data-test="declination-input"
        disabled={false}
        name="declination"
        onChange={[Function]}
      />
    </styled.div>
  </styled.div>
  <styled.div>
    <styled.div>
      <p>
        Search radius
      </p>
      <InputField
        data-test="search-cone-radius-input"
        name="searchConeRadius"
        onChange={[Function]}
      />
    </styled.div>
    <styled.div>
      <p>
        Radius units
      </p>
      <SelectField
        data-test="radius-units-select"
        name="radiusUnits"
        onChange={[Function]}
        value="arcseconds"
      >
        <option
          value="arcseconds"
        >
          Arcseconds
        </option>
        <option
          value="arcminutes"
        >
          Arcminutes
        </option>
        <option
          value="degrees"
        >
          Degrees
        </option>
      </SelectField>
    </styled.div>
  </styled.div>
</Fragment>
`;

exports[`TargetForm should render correctly 2`] = `
<Fragment>
  <styled.div>
    <styled.div>
      <p>
        Target name
      </p>
      <InputField
<<<<<<< HEAD:src/__tests__/components/searchFormComponents/__snapshots__/TargetForm.test.tsx.snap
=======
        className="target-name-input"
>>>>>>> b25246a6a08e483ea3beb3773509373aaefa6819:src/__tests__/components/searchFormComponents/__snapshots__/TargetForm.test.tsx.snap
        data-test="target-name-input"
        disabled={false}
        name="name"
        onChange={[Function]}
        value="NGC 1234"
      />
    </styled.div>
    <styled.div>
      <styled.div>
        <styled.div>
          <p>
            Resolver
          </p>
          <SelectField
<<<<<<< HEAD:src/__tests__/components/searchFormComponents/__snapshots__/TargetForm.test.tsx.snap
=======
            className="resolver-select"
>>>>>>> b25246a6a08e483ea3beb3773509373aaefa6819:src/__tests__/components/searchFormComponents/__snapshots__/TargetForm.test.tsx.snap
            data-test="resolver-select"
            name="resolver"
            onChange={[Function]}
            value="Simbad"
          >
            <option
              value="Simbad"
            >
              Simbad
            </option>
            <option
              value="NED"
            >
              NED
            </option>
            <option
              value="VizieR"
            >
              VizieR
            </option>
          </SelectField>
        </styled.div>
        <styled.div>
          <br />
          <button
            className="button is-info false"
            disabled={false}
            onClick={[Function]}
            title=""
            type="button"
          >
            resolve
          </button>
        </styled.div>
      </styled.div>
    </styled.div>
  </styled.div>
  <styled.div>
    <styled.div>
      <p>
        Right ascension
      </p>
      <InputField
        data-test="right-ascension-input"
        disabled={false}
        name="rightAscension"
        onChange={[Function]}
      />
    </styled.div>
    <styled.div>
      <p>
        Declination
      </p>
      <InputField
        data-test="declination-input"
        disabled={false}
        name="declination"
        onChange={[Function]}
      />
    </styled.div>
  </styled.div>
  <styled.div>
    <styled.div>
      <p>
        Search radius
      </p>
      <InputField
        data-test="search-cone-radius-input"
        name="searchConeRadius"
        onChange={[Function]}
      />
    </styled.div>
    <styled.div>
      <p>
        Radius units
      </p>
      <SelectField
        data-test="radius-units-select"
        name="radiusUnits"
        onChange={[Function]}
        value="arcseconds"
      >
        <option
          value="arcseconds"
        >
          Arcseconds
        </option>
        <option
          value="arcminutes"
        >
          Arcminutes
        </option>
        <option
          value="degrees"
        >
          Degrees
        </option>
      </SelectField>
    </styled.div>
  </styled.div>
</Fragment>
`;

exports[`TargetForm should render correctly 3`] = `
<Fragment>
  <styled.div>
    <styled.div>
      <p>
        Target name
      </p>
      <InputField
<<<<<<< HEAD:src/__tests__/components/searchFormComponents/__snapshots__/TargetForm.test.tsx.snap
=======
        className="target-name-input"
>>>>>>> b25246a6a08e483ea3beb3773509373aaefa6819:src/__tests__/components/searchFormComponents/__snapshots__/TargetForm.test.tsx.snap
        data-test="target-name-input"
        disabled={false}
        error="this is an error"
        name="name"
        onChange={[Function]}
      />
    </styled.div>
    <styled.div>
      <styled.div>
        <styled.div>
          <p>
            Resolver
          </p>
          <SelectField
<<<<<<< HEAD:src/__tests__/components/searchFormComponents/__snapshots__/TargetForm.test.tsx.snap
=======
            className="resolver-select"
>>>>>>> b25246a6a08e483ea3beb3773509373aaefa6819:src/__tests__/components/searchFormComponents/__snapshots__/TargetForm.test.tsx.snap
            data-test="resolver-select"
            name="resolver"
            onChange={[Function]}
            value="Simbad"
          >
            <option
              value="Simbad"
            >
              Simbad
            </option>
            <option
              value="NED"
            >
              NED
            </option>
            <option
              value="VizieR"
            >
              VizieR
            </option>
          </SelectField>
        </styled.div>
        <styled.div>
          <br />
          <button
            className="button is-info false"
            disabled={false}
            onClick={[Function]}
            title=""
            type="button"
          >
            resolve
          </button>
        </styled.div>
      </styled.div>
    </styled.div>
  </styled.div>
  <styled.div>
    <styled.div>
      <p>
        Right ascension
      </p>
      <InputField
        data-test="right-ascension-input"
        disabled={false}
        name="rightAscension"
        onChange={[Function]}
      />
    </styled.div>
    <styled.div>
      <p>
        Declination
      </p>
      <InputField
        data-test="declination-input"
        disabled={false}
        name="declination"
        onChange={[Function]}
      />
    </styled.div>
  </styled.div>
  <styled.div>
    <styled.div>
      <p>
        Search radius
      </p>
      <InputField
        data-test="search-cone-radius-input"
        name="searchConeRadius"
        onChange={[Function]}
        value="12.1"
      />
    </styled.div>
    <styled.div>
      <p>
        Radius units
      </p>
      <SelectField
        data-test="radius-units-select"
        name="radiusUnits"
        onChange={[Function]}
        value="arcminutes"
      >
        <option
          value="arcseconds"
        >
          Arcseconds
        </option>
        <option
          value="arcminutes"
        >
          Arcminutes
        </option>
        <option
          value="degrees"
        >
          Degrees
        </option>
      </SelectField>
    </styled.div>
  </styled.div>
</Fragment>
`;
